### ISSUE TITLE
XrdHTTP PUT does not complain if content-length undefined

### ISSUE THREAD

**Issue by olifre** [xrootd/xrootd#915]
Consider the following "conversation":
```
    => PUT /somewhere/test_webdav_access_9 HTTP/1.1
    => User-Agent: PycURL/7.19.7
    => Host: xrootd006.example.com:1094
    => Accept: */*
    => Transfer-Encoding: chunked
    => Expect: 100-continue
    => 
    <= HTTP/1.1 100 Unknown
    <= Connection: Keep-Alive
    <= Content-Length: 0
    => 79
    => This is a test file generated by the nagios probe check-webdav from package nagios-plugins-webdav. It is safe to delete.
    => 
    => 0
    => 
    <= HTTP/1.1 200 OK
    <= Connection: Keep-Alive
    <= Content-Length: 3
    <= 
    <= :-)
```
The end result is an empty file `/somewhere/test_webdav_access_9` on the storage. 

According to RFC 2616 in (4.4 Message Length):
```
   If a request contains a message-body and a Content-Length is not given,
   the server SHOULD respond with 400 (bad request) if it cannot
   determine the length of the message, or with 411 (length required) if
   it wishes to insist on receiving a valid Content-Length.

   All HTTP/1.1 applications that receive entities MUST accept the
   "chunked" transfer-coding (section 3.6), thus allowing this mechanism
   to be used for messages when the message length cannot be determined
   in advance.
```
So for HTTP 1.1 compatibility, XRootD must accept the content and determine the size on the fly. 

This test was done with XRootD 4.9.0.

**Comment by olifre** [comment]
Rereading the RFC, when "chunked" is used, xrootd should not insist on `Content-Length` nor assume zero length, but learn the length by accepting all data. 
So in fact the communications looks ok - but xrootd only creates a zero-byte file, and it should accept the content.

**Comment by xrootd-dev** [comment]
Hi Oliver,

Yes, you are correct from the protocol standpoint. We are trying to assign 
a priority to this issue. Can you provide a use case where this becomes a 
critical problem?

Andy

**Comment by olifre** [comment]
Hi Andy,

the criticality comes from EGI running the following nagios check against WebDAV endpoints:
https://gitlab.cern.ch/lcgdm/nagios-plugins-webdav/blob/master/src/check_webdav
They plan to make this part of the critical profile (i.e. sties will be marked as broken if they fail the test) sometime in the coming weeks / months (may depend on the number of failures they observe). 

In my opinion, though, the probe is also broken in various ways. It seems to provide a bad `Content-Length` with modern libcurl versions (that's how I found #916 ) and for old libcurl versions, it uses a chunked transfer (but at least in my tests, it sent some garbage numbers as you can see above). 

So in my opinion, the probe issues are worse, but depending on how they are going to fix it, it might be that the fix in xrootd for this issue is still an EGI requirement (even though an FTS does not seem to use chunked encoding, so they test something with the probe that does not match actual real world usage...). 

Cheers,
Oliver

**Comment by bbockelm** [comment]
@olifre - have you determined a way to reproduce this with the CLI?  With RHEL7's curl, I can't get it to do both chunked encoding and `Expect: 100-continue`.  When I enable both, I get:

```
> PUT //tmp/foo HTTP/1.1
> User-Agent: curl/7.29.0
> Host: test.example.com:8000
> Accept: */*
> Transfer-Encoding: chunked
> Expect: 100-continue
> 
< HTTP/1.1 100 Unknown
< Connection: Keep-Alive
< Content-Length: 0
< HTTP/1.1 200 OK
< Connection: Keep-Alive
< Content-Length: 3
< 
* Connection #0 to host test.example.com left intact
:-)
```

@gbitzes - could you help here?

PS - note that chunked encoding sends hex-encoded lengths.  So when @olifre says:

> but at least in my tests, it sent some garbage numbers as you can see above

Things look correct to me - `79` in hex is `121` in decimal, which appears to be the correct length of the data.

**Comment by bbockelm** [comment]
Oh - one thing worth noting is that there appears to be no support for chunked encoding on `PUT`.

So the immediate problem is that it instead of responding with `HTTP/1.1 100 Unknown`, it really ought to be `HTTP/1.1 501 Not Implemented` if the chunked transfer encoding header is present.

**Comment by olifre** [comment]
@bbockelm I did not manage to reproduce via CLI (yet), either. But judging from the comments in:
https://gitlab.cern.ch/lcgdm/nagios-plugins-webdav/blob/0ed87f2f5a54fe0db0adf21216f2741934c5db8c/src/check_webdav#L623
I guess it's probably impossible to reproduce with the CLI alone. They have to put something irrelevant as postfield and hook up a seek function to feed libcurl to make curl do that. 

Thanks for the explanation of the lengths being encoded in hex! Yes, then it's certainly not garbage. Shame on me for not reading the specs completely and good for curl :wink: .

**Comment by gbitzes** [comment]
Hi, I think the following incantation should give the same behavior as the probe - 100-Continue plus chunked encoding:

```
curl https://some-url -k -E /tmp/x509up_u1000 --header "Transfer-Encoding: chunked" --upload-file /etc/services -v -L
```

curl recognizes the added header, and switches automatically to chunked encoding mode.

> (even though an FTS does not seem to use chunked encoding, so they test something with the probe that does not match actual real world usage...).

One of the goals of using HTTP on the grid is allowing the use of generic tools for data access, browsers, curl, etc. Chunked encoding is a very basic feature to support in any HTTP server..

**Comment by olifre** [comment]
@gbitzes Thanks!

> One of the goals of using HTTP on the grid is allowing the use of generic tools for data access, browsers, curl, etc. Chunked encoding is a very basic feature to support in any HTTP server..

I fully agree. However, it would be good to (additionally, not exclusively) test with the same headers which FTS used (i.e. I did not complain the test is bad and should be removed, but it's insufficient to test whether an endpoint is working as expected).

**Comment by bbockelm** [comment]
> One of the goals of using HTTP on the grid is allowing the use of generic tools for data access, browsers, curl, etc. 

Completely agreed!  That's why I put in PR #928 to, in the short term, avoid sending a `100-Continue` and instead respond with a `501 Not Implemented` -- then immediately started working on an implementation itself.  It'll take awhile longer.

That said, if the probe is to be marked critical by EGI, I would like to know which folks critically depend on this feature.  One must decide if it's an RFC compliance test or a site functionality test.

**Comment by bbockelm** [comment]
@olifre - OK, chunked encoding support for `PUT` is included in #931.

Any chance you could test that against the EGI probe?  @gbitzes - are you in a good position to do so?

Would be nice to know we pass the probe prior to the next release....

**Comment by gbitzes** [comment]
@bbockelm Not really, I'm not running XrdHttp...

It's pretty easy to run the probe from the command line, no installation or packages are needed:
```
./check_webdav -vv --uri https://... -E /path/to/proxy
```

**Comment by bbockelm** [comment]
Thanks @gbitzes!

With the posted branch, I can confirm the Nagios probe passes against XrdHttp.

I also had to enable directory listings on GET (which I had turned off previously), which I could see sites complaining about.  I'll let EGI determine the level of criticality there...

**Comment by olifre** [comment]
Thanks @bbockelm ! That's very good to hear, since we still did not manage to free time for a working xroot test setup :cry: 

Do you also have the chance to test the probe on a system with newer libcurl? 
Specifying a bad content length as seen in #916 was interestingly something the probe was doing when executed in an environment with more recent libcurl, and after the fix for #916 in it should now fail (expectedly). 
If the endpoint you used is "open", I can also perform this test from my client with modern libcurl.

**Comment by bbockelm** [comment]
> Do you also have the chance to test the probe on a system with newer libcurl?

No, unfortunately.  I did specify a bad content length by hand in order to fix #916.

> If the endpoint you used is "open"

Unfortunately, it is not open - developer box!

Let's wait for @ffurano to merge this bugfix and I'll try to get an OSG build we can move to our test server.

**Comment by olifre** [comment]
> Unfortunately, it is not open - developer box!

Agreed - I did not really expect you to say "yes", better not to have a dev box open :wink:. 


> Let's wait for @ffurano to merge this bugfix and I'll try to get an OSG build we can move to our test server.

Perfect, let me know once it's available.
